package/samplers/carbo/sampler.py

```diff
@@ -0,0 +1,280 @@
+        is_log[d] = dist.log
+        lows[d] = dist.low
+        highs[d] = dist.high
+    return lows, highs, is_log
+
+
+def _get_params_array(
+    trials: list[FrozenTrial], search_space: dict[str, BaseDistribution]
+) -> np.ndarray:
+    params = np.empty((len(trials), len(search_space)), dtype=float)
+    for i, t in enumerate(trials):
+        for d, (name, dist) in enumerate(search_space.items()):
+            params[i, d] = t.params[name]
+    return params
+
+
+class CARBOSampler(BaseSampler):
+    def __init__(
+        self,
+        *,
+        seed: int | None = None,
+        independent_sampler: BaseSampler | None = None,
+        n_startup_trials: int = 10,
+        deterministic_objective: bool = False,
+        constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
+        rho: float = 1e3,
+        beta: float = 4.0,
+        local_ratio: float = 0.1,
+        n_local_search: int = 10,
+    ) -> None:
+        self._rng = np.random.RandomState(seed)
+        self._independent_sampler = independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        self._intersection_search_space = optuna.search_space.IntersectionSearchSpace()
+        self._n_startup_trials = n_startup_trials
+        self._log_prior: Callable[[KernelParamsTensor], torch.Tensor] = default_log_prior
+        self._minimum_noise = float(DEFAULT_MINIMUM_NOISE_VAR)
+        self._kernel_params_cache: torch.Tensor | None = None
+        self._constraints_kernel_params_cache_list: list[torch.Tensor] | None = None
+        self._deterministic = deterministic_objective
+        self._constraints_func = constraints_func
+        self._beta = beta
+        self._rho = rho
+        assert 0 < local_ratio < 1
+        self._local_ratio = local_ratio
+        self._n_local_search = n_local_search
+
+    def _preproc(
+        self, study: Study, trials: list[FrozenTrial], search_space: dict[str, BaseDistribution]
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        _sign = -1.0 if study.direction == StudyDirection.MINIMIZE else 1.0
+        standardized_score_vals, _, _ = _standardize_values(
+            _sign * np.array([trial.value for trial in trials])
+        )
+        dim = len(search_space)
+        params = _get_params_array(trials, search_space)
+        lows, highs, is_log = _get_dist_info_as_arrays(search_space)
+        X_train = torch.from_numpy(normalize_params(params, is_log, lows, highs))
+        y_train = torch.from_numpy(standardized_score_vals)
+        if self._kernel_params_cache is not None and len(self._kernel_params_cache) - 2 != dim:
+            self._kernel_params_cache = None
+            self._constraints_kernel_params_cache_list = None
+
+        return X_train, y_train
+
+    def _get_warmstart_index(
+        self, X_train: torch.Tensor, y_train: torch.Tensor, constraint_vals: np.ndarray | None
+    ) -> int | None:
+        if constraint_vals is None:
+            return int(y_train[:, 0].argmax().item())
```

Comment:
`y_train` is 1D since CARBO only handles a single objective.
```suggestion
            return int(y_train.argmax().item())
```

Reply:
Thank you for the catch!